content/200-concepts/100-components/02-prisma-client/260-client-extensions/120-query.mdx

```diff
@@ -1,16 +1,168 @@
+    },
+  },
+})
+
+await xprisma.user.findMany() // returns users whose age is greater than 18
+```
+
+In the above example, a call to `xprisma.user.findMany` triggers `query.user.findMany`. Each callback receives a type-safe `{ model, operation, args, data }` object that describes the query. This object has the following properties:
+
+- `model`: the name of the containing model for the query that we want to extend.
+
+  In the above example, the `model` is a string of type `"User"`.
+
+- `operation`: the name of the operation being extended and executed.
+
+  In the above example, the `operation` is a string of type `"findMany"`.
+
+- `args`: the specific query input information to be extended.
+
+  This is a type-safe object that you can mutate before the query happens. You can mutate any of the properties in `args`. Exception: you cannot mutate `include` or `select` because that would change the expected output type and break type safety.
+
+- `data`: a promise for the result of the query.
+
+  - You can use `await` and then mutate the result of this promise, because its value is type-safe. TypeScript catches any unsafe mutations on the object.
+  - `await data` is equivalent to calling `next(...)` in a middleware.
+
+### Add a custom query to a specific operation in all models of your schema
+
+To extend all model queries, use `$allmodels` instead of a specific model name. For example:
+
+```ts
+const xprisma = prisma.$extends({
+  query: {
+    user: {
+      async findMany({ model, operation, args, data }) {
+        // set `age` and fill with the rest of `where`
+        args.where = { age: { gt: 18 }, ...args.where }
+        const users = await data
+        return data
+      },
+    },
+  },
+})
+```
+
+### Add a custom query to all operations in a specific model
+
+Use `$allOperations` to extend all operations at once instead of a specific one.
+
+For example, the following code applies a custom query to all operations on the `user` model:
+
+```ts
+const xprisma = prisma.$extends({
+  query: {
+    user: {
+      $allOperations({ model, operation, args, data }) {
+        // handle all operations
+        return data
+      },
+    },
+  },
+})
+```
+
+### Add a custom query to all operations for all existing models
+
+Use `$allOperations` to apply your custom query to all operations, instead of to a specific operation name.
+
+To apply a custom query to all operations on all models of your schema:
+
+```ts
+const xprisma = prisma.$extends({
+  query: {
+    $allModels: {
+      $allOperations({ model, operation, args, data }) {
+        // handle all operations
+        return data
+      },
+    },
+  },
+})
+```
+
+## Query extensions versus middlewares
+
+You can use query extensions or [middlewares](/concepts/components/prisma-client/middleware#performance-and-appropriate-use-cases) to hook into the query life-cycle and modify an incoming query or its result. Client extensions and middlewares differ in the following ways:
```

Comment:
```suggestion
Besides the `query` Prisma Client Extension component you can also use [middlewares](/concepts/components/prisma-client/middleware#performance-and-appropriate-use-cases) to hook into the query life-cycle and modify an incoming query or its result. Client extensions and middlewares differ in the following ways:
```

Reply:
I prefer my more neutral presentation of this. In the context of the doc, your suggestion flow a little better, but someone might come in at this heading from a search, reading from the context of middleware. For this reason, I think we should leave as-is.